src/cargo/core/compiler/custom_build.rs

```diff
@@ -679,47 +683,120 @@ impl BuildOutput {
+                {
+                    parse_directive(whence.as_str(), line, data, new_syntax, false)?
+                }
+                // For instance, `cargo::metadata=foo=bar`.
+                else if let Some(data) = data.strip_prefix("metadata=") {
+                    // Reserved keys should not be used with the `cargo::metadata=` syntax.
+                    if let Some(prefix) = RESERVED_PREFIXES
+                        .iter()
+                        .find(|prefix| data.starts_with(*prefix))
+                    {
+                        bail!("invalid output in {}: `{}`\n\
+                            The reserved key cannot be used in the `cargo::metadata=` syntax. Please use `cargo::{}VAlUE` instead.\n\
+                            {}", whence, line, prefix, DOCS_LINK_SUGGESTION
+                        )
+                    }
+                    parse_directive(whence.as_str(), line, data, new_syntax, true)?
+                } else {
+                    bail!("invalid output in {}: `{}`\n\
+                        Expected a line with `cargo::metadata=KEY=VALUE` but it did not have the `metadata=` part.\n\
+                        {}", whence, line, DOCS_LINK_SUGGESTION
+                    )
+                }
             }
-            let data = match iter.next() {
-                Some(val) => val,
-                None => continue,
-            };
-
-            // getting the `key=value` part of the line
-            let mut iter = data.splitn(2, '=');
-            let key = iter.next();
-            let value = iter.next();
-            let (key, value) = match (key, value) {
-                (Some(a), Some(b)) => (a, b.trim_end()),
-                // Line started with `cargo:` but didn't match `key=value`.
-                _ => bail!("invalid output in {}: `{}`\n\
-                    Expected a line with `cargo:key=value` with an `=` character, \
-                    but none was found.\n\
-                    See https://doc.rust-lang.org/cargo/reference/build-scripts.html#outputs-of-the-build-script \
-                    for more information about build script outputs.", whence, line),
+            // For backwards compatibility, we also accept lines that start with "cargo:".
+            else if let Some(data) = line.strip_prefix("cargo:") {
+                // For instance, `cargo:rustc-flags=foo` or `cargo:foo=bar`.
+                parse_directive(whence.as_str(), line, data, false, false)?
+            }
+            // Skip this line since it doesn't start with "cargo:" or "cargo::".
+            else {
+                continue;
             };
 
             // This will rewrite paths if the target directory has been moved.
             let value = value.replace(
                 script_out_dir_when_generated.to_str().unwrap(),
                 script_out_dir.to_str().unwrap(),
             );
-
+            let syntax_prefix = if new_syntax { "cargo::" } else { "cargo:" };
```

Comment:
Do we need to bother with this?

Reply:
I'm not sure. Wouldn't it be unfriendly or strange if the user was using the old syntax, but the error message was the new one?

Reply:
@weihanglo what are your thoughts on whether its worth the complexity to track the syntax used to adapt error messages for the user?

Reply:
I slightly agree with hi-rustin.

Maybe, instead of having repeatitive if-else and `cargo:`, we could have an enum to reduce the boilerplate.

```rust
// name bikeshedding welcome
enum Prefix {
    Old,
    Current,
}

impl Prefix {
    fn as_str(&self) -> &str {
        match self {
            Self::Old => "cargo:",
            Self::Current => "cargo::",
        }
    }
}
```

But no. I am making it more complicated 😆.